Coverage reporting and setup.py fixes

thomasw opened the pull request:
I initially wanted to just add coverage reporting, but I ended up needing to take care of a number of additional setup.py changes to support of that. This pull:
1. Configures coverage reporting and readies the repo for https://coveralls.io/
2. Fixes setup.py so that all of its commands work correctly without manually installing things from requirements.txt
3. Installs and configures dependencies to make building signed universal wheel dists reasonable (https://hynek.me/articles/sharing-your-labor-of-love-pypi-quick-and-dirty/)
4. Reconfigures travis so that it executes the test suite via setup.py for added verification. Note that it still uses pip to install requirements in that environment so that we can take advantage of build dependency caching.

thomasw committed:
gitignore: Ignore coverage reports and egg related files

When we toggle on coverage reporting, .coverage files will start
being generated on tests runs. Additionally, we don't want .egg
files to dirty the repo state either if we verify the setup.py
file by running `python setup.py nosetests`.

thomasw committed:
Install coverage related requirements and configure them

thomasw committed:
.travis.yml: whitespace cleanup

thomasw committed:
travis: configure travis to cache pip installs between builds

thomasw committed:
travis: push coverage reports to coveralls after test runs

thomasw committed:
travis: configure travis so that it runs tests with setup.py

Running nosetests via setup.py adds some small amount of setup.py
verification to your tests runs for free. If setup.py is broken
for some reason, your tests will blow up. This is fairly opinionated
but it's saved me in the past before.

thomasw committed:
requirements: only require mock for python < 3.3

Use an environment marker based requirements specifier so that we don't install mock in environments where it isn't necessary.

mock was added to the stdlib in python 3.3: https://docs.python.org/3/library/unittest.mock.html

https://pip.pypa.io/en/stable/reference/pip_install/#requirement-specifiers

thomasw committed:
Configure nosetests to ignore tube.py

Because we're using the doctest plugin, nosetests will look in tube.py for docs with tests in them. That's okay in tdubs dev environments because testtube will be installed. However, when running python setup.py nosetests in a non-dev environment context, the tests command will fail because we don't list testtube in setup_requires.

thomasw committed:
setup.py: specify dependencies for all commands

In order for all of setup.py's commands to run correctly, we need setuptools to first install several dependencies if they aren't already present. Without these things installed, 1. the test command won't execute any tests. 2. the nosetests command won't be found if you try to use setup.py to run it and it will throw error if you try to execute it.

After these changes, one should be able to download the tdubs package and verify it in an empty virtual environment.

thomasw committed:
manifest: include the changelog in dists

thomasw committed:
manifest: include test suite in dists

This enables users to verify the dists downloaded from pypi using setup.py

thomasw committed:
Add changelog entries

thomasw committed:
Add twine and wheel to our requirements file

tdubs should be uploaded as a universal wheel to pypi. The twine and wheel will allow the maintainer to do that. These dependencies will also enable the maintainer to sign the package for added verification.

See https://hynek.me/articles/sharing-your-labor-of-love-pypi-quick-and-dirty/ for full instructions.

thomasw committed:
Fix for setup.py multiprocessing issue

In python2.6, using nose via setup.py breaks if multiprocessing isn't imported. This change fixes that problem for that environment and is otherwise a noop.

thomasw committed:
Don't install unsupported depdendencies in 3.2

Coverage dependencies are unsupported in python 3.2 and cause things to break when we attempt to install them. This disables them for that environment specifically, however it might be more prudent to drop support for 3.2 as a number of other packages have done already.

thomasw commented:
I'm pretty sure the coverage reporting that I added is broken because it doesn't appear to be taking doctests into account. I could've sworn I was seeing 100% coverage early on in this branch, but then more recently I've only been seeing 64%. It still might be worth turning the reports on and then eventually working on that problem in isolation.

thomasw committed:
Add coverage reporting for python 3.2

For python 3.2, we need to use an older version of coverage that has 3.2 support.

thomasw committed:
Make coverage config more reselient

Previously, we were telling coverage to report on tdubs.py specifically, but this would've broken if that implementation was ever moved into a folder based module. This new configuration should support that without additional changes.

blaix commented on setup.cfg:
```diff
@@ -2,3 +2,9 @@
 doctest-options=+ELLIPSIS,+IGNORE_EXCEPTION_DETAIL
 doctest-extension=rst
+with-coverage=1
+ignore-files=tube.py
```
?

blaix replied:
I'm assuming this is because I'm using a top-level module instead of a package.

blaix replied:
Nope. Not sure what's going on here. When I run `nosetests` on this branch I get a strange error about a missing command. But if I remove this line, all the tests run and I see 100% coverage.

thomasw replied:
This line is because: https://github.com/blaix/tdubs/pull/14/commits/18d743e1a30f53b580fbb78e9c74e911be9d6ce2

blaix replied:
Wow, I'm dumb. I saw `tube.py` but read `tdubs.py`.

Duh.

blaix commented:
> I'm pretty sure the coverage reporting that I added is broken because it doesn't appear to be taking doctests into account.

That's because you're running them with setup.py isn't it? Some googling said we may be able to get setup.py to run them with a custom setuptools Command subclass, but do you think that's worth the effort for the extra verification of setup.py?

blaix commented:
>  It still might be worth turning the reports on and then eventually working on that problem in isolation.

I'd rather remove the travis/setup.py change from this pull, assuming that's the problem, and then change it back in another pull alongside a fix that gets setup running the doctests.

blaix commented:
By the way, you're a super hero for doing all this. Thanks!

thomasw commented:
> That's because you're running them with setup.py isn't it? Some googling said we may be able to get setup.py to run them with a custom setuptools Command subclass, but do you think that's worth the effort for the extra verification of setup.py?

No, I think you're missing some context here. Your test suite is currently dependent on nose for proper discovery. `python setup.py test` is never going to run your full test suite because, even though I've configured it to use nose for discovery (https://github.com/blaix/tdubs/pull/14/files#diff-2eeaed663bd0d25b7e608891384b7298R49), doctest, coverage, and other setup.cfg configured nose plugins aren't supported via the test command. You'll see errors relating to each of those plugins, but the test command won't fail. The nose docs recommend that you simply abandon `python setup.py test` and use `python setup.py nosetests` instead, which has full plugin support. Most of the changes I've done here are in support of `python setup.py nosetests`, not `python setup.py test` (which now does something at least whereas before it did nothing).

The bad coverage report data that I was seeing was when running both `python setup.py nosetests` or `nosetests` (after pip installing requirements).

Note: to properly test that setup.py's test and nosetests commands work, you need to run them from an empty virtual environment.

An alternative to what I've done here is to make it so that you aren't dependent on nose, but I think you'd end up losing valuable things if you went down that road (like each doctest showing up as a separate test result).

thomasw committed:
Correctly configure nose's ignore-files option

When one specifies an ignore-files configuration for nose, it clobbers the default config. This is apparently by design and is discussed here: https://github.com/nose-devs/nose/issues/872

One of the things the default config does is ignore setup.py. Not ignoring it causes really weird behavior including breaking the `nosetests` command and messing up coverage reports. This change amends the default ignore-files config so that it also ignores tube.py.
